config/deploy.yml

```diff
@@ -0,0 +1,108 @@
+# Initially this file has some FIXMEs that you need to fill out to get started
+# with Kamal.
+#
+# This deploy.yml assumes the following prerequisites:
+#   - You have a Docker Hub account with a `docker-hub-username`.
+#   - You have a server you can SSH into.
+#   - You have a domain name that you own and will point to your server's IP address.
+#
+# Name of your application. Used to uniquely configure containers.
+service: untitled-application
+
+# Name of the container image.
+# FIXME: Change this to your Docker Hub username.
+image: docker-hub-username/untitled-application
+
+# Deploy to these servers.
+servers:
+  web:
+    - 999.999.999.999 # FIXME: The IP address of the server you can SSH into.
+  job:
+    hosts:
+      - 999.999.999.999 # FIXME: The IP address of the server you can SSH into.
+    cmd: bundle exec sidekiq
+
+# Enable SSL auto certification via Let's Encrypt and allow for multiple apps on
+# a single web server. Remove this section when using multiple web servers and
+# ensure you terminate SSL at your load balancer.
+#
+# Note: If using Cloudflare, set encryption mode in SSL/TLS setting to "Full" to
+# enable CF-to-app encryption.
+proxy:
+  ssl: true
+  host: app.yoursaas-fix-me.com # FIXME: The domain you want to use for your app. You need to create a CNAME record that points to the IP address of the server you can SSH into.
+  app_port: 3000
+
+# Credentials for your image host.
+registry:
+  # Specify the registry server, if you're not using Docker Hub
+  # server: registry.digitalocean.com / ghcr.io / ...
+  username: docker-hub-username # FIXME: The username you use to log into Docker Hub.
+  # Always use an access token rather than real password (pulled from .kamal/secrets).
+  password:
+    - KAMAL_REGISTRY_PASSWORD
+
+# Configure builder setup.
+builder:
+  # By default, Kamal will only deploy committed changes. During development,
+  # you may want to uncomment this line to deploy uncommitted changes as well.
+  # context: .
+  arch: arm64 # FIXME: Match your server's architecture.
+
+# Inject ENV variables into containers (secrets come from .kamal/secrets).
+env:
+  secret:
+    - RAILS_MASTER_KEY
+    - POSTMARK_API_TOKEN # FIXME: Keep it or remove it and change the email provider. Keeping it will enable Postmark for email delivery.
+    - UNTITLED_APPLICATION_DATABASE_PASSWORD
+    - REDIS_PASSWORD
+    - POSTGRES_PASSWORD
+  clear:
+    DB_HOST: untitled-application-db
+    POSTGRES_USER: untitled_application
+    POSTGRES_DB: untitled_application_production
+    DB_PORT: 5432
+    REDIS_HOST: untitled-application-redis
+    REDIS_PORT: 6379
```

Comment:
I think some of these are likely to be things that you'd want to be different between staging and production (for example) and so we may not want to encourage them to be hard coded. And if we go with `REDIS_URL` and `DATABASE_URL` then a lot of these should just go away.

Reply:
I'm not sure we'll be able to replace all of them with their `_URL` counterparts. _I think_ some of these vars could be default names that are used inside the containers.

Reply:
It might be fine to "hardcode" them here, because you'd typically create something like a new `deploy.staging.yml` to deploy to a different env. 🤔

Reply:
Now I remember, the problem I'm running into with these is that the URLs are "secrets" because they contain the password. So you'd define them in your local `.zshrc` (or similar) or 1Password, but those places have no knowledge of ports, hosts, etc.

```bash
export REDIS_URL="redis://:${REDIS_PASSWORD}@${REDIS_HOST}:${REDIS_PORT}/0"
export DATABASE_URL="postgres://${POSTGRES_USER}:${POSTGRES_PASSWORD}@${DB_HOST}:${DB_PORT}/${POSTGRES_DB}"
```

Then we'd have a bad type of duplication, where we'd need to fix things in both places (risking to forget it in one of them).

Reply:
We could probably put all of that into the secret env 🤔

Reply:
Oof, I see what you mean. 🤔